### PR TITLE
added KMP linear pattern searching algorithm

### PR DESCRIPTION
**Pull request by sashodada**
Айде да пробваме с PR-и как ше се справим 😆

### CODE REVIEW

File: patterns.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    std::vector<int> pattern_search_kmp(const char *text, const char *pattern) {
+        std::vector<int> result;
+
+        auto lps = prepare_lps(pattern);
+        int pattern_length = strlen(pattern);
+        int i = 0;
+        int j = 0;
+        while (i < strlen(text)) {
```
Comment (GeorgiSamardzhiev) [inline patterns.cpp]: Shouldn't strlen(text) be outside of the while condition?

File: patterns.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
+#include <vector>
```
Comment (GeorgiSamardzhiev) [inline patterns.cpp]: Should separete in hpp and cpp files and add #ifndef #define part, so one could include the file more than once.

**Review by GeorgiSamardzhiev — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+*.o
```
Comment (GeorgiSamardzhiev) [inline .gitignore]: There will be a bigger gitignore file in future commits we don't need that now.

**Review by GeorgiSamardzhiev — CHANGES_REQUESTED** [review]
I will add big gitignore file.

**Comment by sashodada** [comment]
ша мърджаме ли 🗡?

**Pull request merged**